src/bytes/complete.rs

```diff
@@ -242,55 +243,27 @@ where
 {
   move |i: I| {
     let input = i;
+    let input_len = input.input_len();
+
+    // get m and n in bytes
+    let m = match input.slice_index(m) {
```

Comment:
calling slice_index twice before even processing the predicate can be very inefficient for strings, since it has to iterate over chars. Do you see another way to do it?

Reply:
Hm. We have to get the byte offsets for `m` and `n` before we can return any result and it doesn't really make a difference if we do that before or after matching the pattern, right? The only alternative I can think of is changing `Input::position()` to return a byte offset for `&[u8]` and a char offset for `&[str]`, then we are consistently using either byte or char offsets in `take_while_m_n` and can skip those calls to `Input::slice_index()`. But I'm not sure how much that would break.

Reply:
none of those sound like good alternatives. every "split" function should only work with byte offsets, but we cannot compare those with m and n. I feel like the best approach would be to reimplement manually with the `iter_elements` method, which for `&str` calls into `char_indices`, instead of using `position`. Like, iter while the condition is satisfied, stop after n iterations or when the condition fails, then check that the number of iterations was higher than m

Reply:
If it helps, winnow-rs/winnow#130 is how I fixed this though there will be slight differences due to the design of `Input` vs `Stream`